### ISSUE TITLE
highlight page numbers are dependent on font/screen size

### ISSUE THREAD

**Issue by devvythelopper** [koreader/koreader#12153]
* KOReader version: 2023.04
* Device: linux

#### Issue

Highlights are stored the page number currently shown on the screen, which is dependent on font-size and window/screen size. If you then export your highlight, those same page numbers are used. This can result in very confusing highlight references in case you changed font size while reading. Please implement device independent page numbers for highlight exports.

For example by rendering the book when exporting highlights in a specific default size, that will never change, and making the page numbers dependent to that. 

edit: It would be even better if the implementation also supported epub page-maps where available.

**Comment by hius07** [comment]
Can you elaborate? Assume you made a highlight in page 5.
You see page 5 in the status bar.
(1) What page number do you want to see in the Bookmark list for this highlight?
(2) What page number do you want to see in the exported highlight list?
Then you change the font size and the highlight is in the page 6 (page number in the status bar).
The same two questions, (2) for newly exported list obviously.

**Comment by devvythelopper** [comment]
Yeah, so (hypothetically) I made a highlight on page 5. When I export right now, it will say that the highlight is on page 5. Then I change font-size or resize the window. Now my highlight appears on page 10. I put another highlight right on the phrase before the one I just highlighted. Now I export my highlights. The export will contain the highlight which it claims was on "page 5" and afterwards the highlight on "page 10". 

But of course the two highlights should both be on the same page, and the highlight that is exported as "page 10" is the phrase directly before the one on "page 5". 

I don't care whether "page 10" or "page 5" is what is being exported. But I do want the highlights exported in the right order and to have two things that are actually beside each other appear on either the same page or consecutive pages.

I think a standard page and font size that is computed for exports would make sense. Because I use Koreader also on two different sized e-ink devices. And if the export depends on the current font size and screen size of whichever device I use, I won't be able to compare the result of the one from my Oasis with the one from my paperwhite or the one from my linux machine. I mean I can do the export myself and set a specific font and screen size whenever I do, but would that truly be a desireable state of affairs?

**Comment by devvythelopper** [comment]
@hius07 Yeah, so (hypothetically) I made a highlight on page 5. When I export right now, it will say that the highlight is on page 5. Then I change font-size or resize the window. Now my highlight appears on page 10. I put another highlight right on the phrase before the one I just highlighted. Now I export my highlights. The export will contain the highlight which it claims was on "page 5" and afterwards the highlight on "page 10".

But of course the two highlights should both be on the same page, and the highlight that is exported as "page 10" is the phrase directly before the one on "page 5".

It does not matter whether "page 10" or "page 5" is what is being exported. But I do want the highlights exported in the right order and to have two things that are actually beside each other appear on either the same page or consecutive pages.

I think a standard page and font size that is computed for exports would make sense. Because I use Koreader also on two different sized e-ink devices. And if the export depends on the current font size and screen size of whichever device I use, I won't be able to compare the result of the one from my Oasis with the one from my paperwhite or the one from my linux machine. I mean I can do the export myself and set a specific font and screen size whenever I do, but would that truly be a desireable state of affairs?

**Comment by hius07** [comment]
> The export will contain the highlight which it claims (in the export!) is on "page 5" and afterwards the highlight on "page 10".

It is fixed in the current nightly build. Export now shows correct page numbers.

**Comment by hius07** [comment]
> I think a standard page and font size that is computed for exports would make sense.

Define "standard" please.

**Comment by hius07** [comment]
Anyway, sync/export/import of highlights between devices is one of the most frequent FRs, we should think over it.

**Comment by poire-z** [comment]
My thoughts (previously shared) is that sync/import ends up with highlights ending up on a device/computer with KOReader, where they will just work and get you to the page. The page number on the right has no real importance, except approximately telling where in the book they are.
It is only when exported and viewed in a browser or printed on paper (not the _most frequent_ thing mentionned) that "randomely scaled" page numbers may be bothering and hard to use. But is there really some use ? Most often, you'll want to read the text printed on paper. Rarely you need to get back to the device, and if you do, you can find the highlight in the Bookmarks list - a page number won't help much, even more if you can't jump to it because it doesn't match the current screen pages. (May be a highlight sequence number could help finding it again in the Bookmarks list page?)
It is only when wanting to find it again elsewhere than in KOReader (ie. paperbook) where page number would help - and only if they are reference pages, which I think is something that can be exported.

(Or may be you should save 2 "page numbers" in an annotation: the screen page number, and a fixed page number - reference page or synthetic page number - and have a toggle in the Bookmarks list widget. I think we discussed that elsewhere.)

There ends my _should think it over_ :)

**Comment by mergen3107** [comment]
> The page number on the right has no real importance, except approximately telling where in the book they are.

How about replacing “page number” with location percentage? This is common between devices no matter what scale and number of pages is used.

**Comment by jonnyl2** [comment]
Percentages work well for similarly sized, smaller documents but are less user-friendly for larger documents. Many epubs exist with huge collections of classic works; often well over 10,000 rendered pages. Then, it is hard to estimate how many pages are between a highlight at 34.7% and one at 34.9%, for example. (And 0.1% would spread over 10 or more pages in this example.)

My personal favorite would be reference pages based on character count, for which a facility already exists within KOReader (thanks to @poire-z I believe). Especially if the count is user-definable in the UI, so it can be set according to individual preference.

**Comment by poire-z** [comment]
Added in https://github.com/koreader/crengine/pull/478, and still being discussed in https://github.com/koreader/koreader/issues/9020#issuecomment-1119588295.
Still some open questions, ie. https://github.com/koreader/koreader/issues/9020#issuecomment-2045865939.
Better keep the discussion there.

**Comment by benoit-pierre** [comment]
Store them as percentage, or a normalized 0.0-1.0 floating point, and calculate the page number at display time (according to the book current settings).

**Comment by devvythelopper** [comment]
from @hius07 
> > The export will contain the highlight which it claims (in the export!) is on "page 5" and afterwards the highlight on "page 10".
> 
> It is fixed in the current nightly build. Export now shows correct page numbers.

great, thanks!

> > I think a standard page and font size that is computed for exports would make sense.
> 
> Define "standard" please.

This is a question for the experts, weighing pros and cons. The main point is, that *any* standard would be better than device- or setting-dependent exports. It might even be implemented by giving user the option to choose from a set of standards (i.e. choosing a print book size, and its fine if the page numbers that result are only rough correlates in the absence of a real page-map).

from @jonnyl2 

> The page number on the right has no real importance, except approximately telling where in the book they are. It is only when exported and viewed in a browser or in KOHighlight or printed on paper (not the _most frequent_ thing mentionned) that "randomely scaled" page numbers may be bothering and hard to use. But is there really some use ? Most often, you'll want to read the text printed on paper. Rarely you need to get back to the device, and if you do, you can find the highlight in the Bookmarks list - a page number won't help much, even more if you can't jump to it because it doesn't match the current screen pages. (May be a highlight sequence number could help finding it again in the Bookmarks list page?) It is only when wanting to find it again elsewhere than in KOReader (ie. paperbook) where page number would help - and only if they are reference pages, which I think is something that can be exported.

my use case is this: I am reading ebooks for academic and semi-academic purposes. Hence I must create actually usable references that people can interpret. I of course make a lot more highlights in a book, than I will actually reference directly. So it is fine for me to look up highlights I will actually use in a printed book (page-maps would of course be *ideal*). And hence it is important for me to be able to interpret the references myself in such a way, that I can find them in a hardcopy or any another edition of the book.

So if I could for the export select the book size I have in hand, and that would give me rough page estimates that would be great!

With that in mind, it would also be helpful if the export did contain the chapter the reference is in, and maybe a percentage of where the highlight is located in the chapter. 

In fact it would be great if the export contained the whole structure of the book (all headlines and subheadlines), so that the highlights are given context.

**Comment by jonnyl2** [comment]
^ that's @poire-z's quote; just for the record :)

**Comment by hius07** [comment]
Fixed by synthetic pages and `pageref` in the export.
